calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,380 @@
+            split_analysis.snd_asmts,
+            builder,
+            format!("end_spl_{}", group_name.id),
+        );
+
+        Ok((first_group, second_group))
+    }
+
+    // Goes through assignments, and properly fills in the fields go_done_asmt,
+    // first_go_asmt, fst_asmts, snd_asmts, and group_done_asmt.
+    fn organize_assignments(
+        &mut self,
+        mut assigns: Vec<ir::Assignment>,
+        first_cell_name: &ir::Id,
+        second_cell_name: &ir::Id,
+    ) {
+        for asmt in assigns.drain(..) {
+            match writes_to_cell(&asmt) {
+                Some(cell_name) => {
+                    if Self::is_go_done(&asmt) {
+                        self.go_done_asmt = Some(asmt);
+                    } else if Self::is_specific_go(&asmt, first_cell_name) {
+                        self.first_go_asmt = Some(asmt);
+                    } else if cell_name == first_cell_name {
+                        self.fst_asmts.push(asmt);
+                    } else if cell_name == second_cell_name {
+                        self.snd_asmts.push(asmt);
+                    } else {
+                        unreachable!(
+                            "Does not write to one of the two \"stateful\" cells"
+                            )
+                    }
+                }
+                None => self.group_done_asmt = Some(asmt),
+            }
+        }
+    }
+
+    // Builds ordering for self. Returns true if this is a complete, valid,
+    // linear ordering in which all reads from the fist cell are from a
+    // stable port, false otherwise.
+    pub fn possible_split(
+        &mut self,
+        asmts: &[ir::Assignment],
+    ) -> Option<(ir::Id, ir::Id)> {
+        let v = ReadWriteSet::write_set(asmts.iter())
+            .map(|cell| cell.clone_name())
+            .collect::<Vec<ir::Id>>();
+
+        if v.len() == 2 {
+            //Update self.go_done_map and self.last for each asmt in the group.
+            for asmt in asmts {
+                self.update(asmt);
+            }
+            if let (Some(last), Some((maybe_first, maybe_last))) =
+                (self.last.clone(), self.done_go.clone())
+            {
+                if maybe_last == last
+                // making sure maybe_fist and maybe_last are the only 2 cells written to 
+                && v.contains(&maybe_first)
+                && v.contains(&maybe_last)
+                // making sure that all reads of the first cell are from stable ports 
+                && asmts.iter().all(|assign| {
+                    if_name_stable_or_done(assign, &maybe_first)
+                }) {
+                    return Some((maybe_first, maybe_last));
+                }
+            }
+        }
+        None
+    }
+
+    // For a given asmt, if asmt is a.go = b.done, then we add (b,a) to self.go_done_map.
+    // Also if asmt is group[done] = cell.done, sets self.last to Some(cell).
+    fn update(&mut self, asmt: &ir::Assignment) {
```

Comment:
Make this function return `Some((last, go_done))` instead of updating the fields on SplitAnalysis